lib/account/account-manager.js

```diff
@@ -83,6 +145,147 @@ module.exports = function (accountService, config, hotp, otDate, promise, random
+            return accountService.getAccountFileAsync(loginFile.accountId).then((accountFile) => {
+                return loginVerifyAsync(accountFile, loginFile.salt, loginInfo.password);
+            }).then(() => {
+                checkMfa(loginFile.mfaKey, loginInfo.currentMfa);
+
+                return accountService.deleteLoginFileAsync(hashedAccountId, loginFile.challengeId);
+            });
+        });
+    }
+
+
+    /**
+     * Need to verify the password we have on file matches
+     * what the account holder is using to log in.
+     *
+     * @param {accoutManager~accountFile} accountFile
+     * @param {string} salt
+     * @param {string} hashedPassword
+     * @return {boolean}
+     * @throws {Error} If the password hashes do not match
+     */
+    function loginVerifyAsync(accountFile, salt, hashedPassword) {
+        return hashPasswordAsync(accountFile, salt).then((passwordHash) => {
+            if (! secureHash.compare(passwordHash, base64.encode(hashedPassword))) {
+                throw new Error("Password hashes do not match");
+            }
+
+            return true;
+        });
+    }
+
+
+    /**
+     * Maps information we want from the service to return
+     * to the account holder to continue the login process
+     *
+     * @param {accountManager~loginInitiateDetails} details
+     * @return {accountManager~loginInitiateClientInfo}
+     */
+    function loginInitiateDetails(details) {
+        return accountService.getAccountFileAsync(details.accountId).then((accountFile) => {
+            return random.randomIdAsync(config.account.passwordSaltLength).then((passwordSaltId) => {
+                var passwordHash;
+
+                passwordHash = accountConfig.passwordHash;
+                passwordHash.salt = accountFile.passwordSalt;
+
+                return {
+                    pbkdf: passwordHash,
+                    challenge: {
+                        salt: details.salt,
+                        hash: accountConfig.loginHash
+                    },
+                    encoding: "base64"
+                };
+            });
+        });
+    }
+
+
+    /**
+     * Initiates the login process by creating the challenge id
+     * and a salt used for logging in.
+     *
+     * @param {accountManager~loginInitateInfo} loginInitiateInfo
+     * @return {accountManager~loginInitateDetails}
+     * @throws {Error} When account hashes do not match
+     */
+    function loginInitiationAsync(hashedAccountId, loginInitiateInfo) {
+        return secureHash.hashAsync(loginInitiateInfo.accountId, accountConfig.idHash).then((accountIdHash) => {
+
+            if (! secureHash.compare(accountIdHash, hashedAccountId)) {
+                throw new Error("Account hashes do not match");
+            }
+
+            return promise.props({
+                challengeId: random.randomIdAsync(accountConfig.loginIdLength),
+                salt: random.randomIdAsync(accountConfig.passwordSaltLength)
```

Comment:
What is the difference between challengeId and salt?  It seems to me that they are really the same thing?